### ISSUE TITLE
[MAC] Convolution with kernel size 3 yields different results depending on whether gradient is enabled or not.

### ISSUE THREAD

**Issue by FelipeMoser** [pytorch/pytorch#128945]
### 🐛 Describe the bug

```python
import torch as tt
import torch.nn as nn

# Random image
image = tt.randn(1, 1, 512, 512)

# Conv2d with kernel size of 3 results in different output if conv_layer has gradients enabled or not
conv_layer_3 = nn.Conv2d(1, 3, 3)
test_3_a = conv_layer_3(image)
conv_layer_3.requires_grad_(False)
test_3_b = conv_layer_3(image)
print(tt.equal(test_3_a, test_3_b)) # False. The differences are between 1e-5 and 1e-8

# Conv2d with kernel size of 5 does not show the same behaviour
conv_layer_5 = nn.Conv2d(1, 3, 5)
test_5_a = conv_layer_5(image)
conv_layer_5.requires_grad_(False)
test_5_b = conv_layer_5(image)
print(tt.equal(test_5_a, test_5_b)) # True
```

### Versions

[pip3] numpy==1.26.4
[pip3] torch==2.5.0.dev20240617
[pip3] torchaudio==2.4.0.dev20240617
[pip3] torchvision==0.19.0.dev20240617
[conda] nomkl                     1.0                  h5ca1d4c_0    conda-forge
[conda] numpy                     1.26.4          py312h8442bc7_0    conda-forge
[conda] pytorch                   2.5.0.dev20240617        py3.12_0    pytorch-nightly
[conda] torchaudio                2.4.0.dev20240617       py312_cpu    pytorch-nightly
[conda] torchvision               0.19.0.dev20240617       py312_cpu    pytorch-nightly

cc @malfet @albanD

**Comment by malfet** [comment]
Hmm, the differences I'm seeing are in range of 1e-7, but it is a bit strange that they are there to begin with

**Comment by FelipeMoser** [comment]
The range changes depending on the input data. For the data I'm working with I get 1e-5 but for these random minimal examples I get 1e-7 and 1e-8 values.

**Comment by NiharJani2002** [comment]
Hii @malfet  and @FelipeMoser . I am also a macOS user and experienced the same things, I further experimented with and found out two things:-)
1) I experiment from 1 to 5 kernel_size and found out 3 is the only problem.
Code:-)
***
import torch as tt
import torch.nn as nn

# Random image
image = tt.randn(1, 1, 512, 512)

# Conv2d with kernel size of 1 results in different output if conv_layer has gradients enabled or not
conv_layer_1 = nn.Conv2d(1, 1, 1)
test_1_a = conv_layer_1(image)
conv_layer_1.requires_grad_(False)
test_1_b = conv_layer_1(image)
print(tt.equal(test_1_a, test_1_b))  # True


# Conv2d with kernel size of 2 results in different output if conv_layer has gradients enabled or not
conv_layer_2 = nn.Conv2d(1, 2, 2)
test_2_a = conv_layer_2(image)
conv_layer_2.requires_grad_(False)
test_2_b = conv_layer_2(image)
print(tt.equal(test_2_a, test_2_b))  # True

# Conv2d with kernel size of 3 results in different output if conv_layer has gradients enabled or not
conv_layer_3 = nn.Conv2d(1, 3, 3)
test_3_a = conv_layer_3(image)
conv_layer_3.requires_grad_(False)
test_3_b = conv_layer_3(image)
print(tt.equal(test_3_a, test_3_b)) # False. The differences are between 1e-5 and 1e-8

# Conv2d with kernel size of 4 results in different output if conv_layer has gradients enabled or not
conv_layer_4 = nn.Conv2d(1, 4, 4)
test_4_a = conv_layer_4(image)
conv_layer_4.requires_grad_(False)
test_4_b = conv_layer_4(image)
print(tt.equal(test_4_a, test_4_b))  # True

# Conv2d with kernel size of 5 does not show the same behavior
conv_layer_5 = nn.Conv2d(1, 5, 5)
test_5_a = conv_layer_5(image)
conv_layer_5.requires_grad_(False)
test_5_b = conv_layer_5(image)
print(tt.equal(test_5_a, test_5_b)) # True
***
Output:
***
True
True
False
True
True
***
<img width="1216" alt="image" src="https://github.com/pytorch/pytorch/assets/84491997/d3520ae4-a359-41cb-87dc-48dba5a48593">


2. Second thing I noticed is when I keep the precision to float64, then all are True.
***
Code:
import torch as tt
import torch.nn as nn

# Random image
image = tt.randn(1, 1, 512, 512).type(tt.float64)

# Conv2d with kernel size of 1 results in different output if conv_layer has gradients enabled or not
conv_layer_1 = nn.Conv2d(1, 1, 1,dtype=tt.float64)
test_1_a = conv_layer_1(image)
conv_layer_1.requires_grad_(False)
test_1_b = conv_layer_1(image)
print(tt.equal(test_1_a, test_1_b))  # True


# Conv2d with kernel size of 2 results in different output if conv_layer has gradients enabled or not
conv_layer_2 = nn.Conv2d(1, 2, 2,dtype=tt.float64)
test_2_a = conv_layer_2(image)
conv_layer_2.requires_grad_(False)
test_2_b = conv_layer_2(image)
print(tt.equal(test_2_a, test_2_b))  # True

# Conv2d with kernel size of 3 results in different output if conv_layer has gradients enabled or not
conv_layer_3 = nn.Conv2d(1, 3, 3,dtype=tt.float64)
test_3_a = conv_layer_3(image)
conv_layer_3.requires_grad_(False)
test_3_b = conv_layer_3(image)
print(tt.equal(test_3_a, test_3_b)) # False. The differences are between 1e-5 and 1e-8

# Conv2d with kernel size of 4 results in different output if conv_layer has gradients enabled or not
conv_layer_4 = nn.Conv2d(1, 4, 4,dtype=tt.float64)
test_4_a = conv_layer_4(image)
conv_layer_4.requires_grad_(False)
test_4_b = conv_layer_4(image)
print(tt.equal(test_4_a, test_4_b))  # True

# Conv2d with kernel size of 5 does not show the same behavior
conv_layer_5 = nn.Conv2d(1, 3, 5,dtype=tt.float64)
test_5_a = conv_layer_5(image)
conv_layer_5.requires_grad_(False)
test_5_b = conv_layer_5(image)
print(tt.equal(test_5_a, test_5_b)) # True
***
Output:
True
True
True
True
True

<img width="1216" alt="image" src="https://github.com/pytorch/pytorch/assets/84491997/2a26dc80-c842-4164-b5ae-11845c3e2f27">

**Comment by janeyx99** [comment]
Following discussion with the group: different kernels could be used depending on whether grad is turned on or off, so minor numerical discrepancies are too be expected.

**Comment by janeyx99** [comment]
Closing this issue as expected behavior, but feel free to call out if these discrepancies are more than what is expected from https://pytorch.org/docs/stable/notes/numerical_accuracy.html#numerical-accuracy
